src/Grpc.AspNetCore.Web/Internal/Base64PipeReader.cs

```diff
@@ -0,0 +1,222 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using Microsoft.AspNetCore.Authorization.Infrastructure;
+using System;
+using System.Buffers;
+using System.Buffers.Text;
+using System.Diagnostics;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Grpc.AspNetCore.Web.Internal
+{
+    /// <summary>
+    /// Reads and decodes base64 encoded bytes from the inner reader.
+    /// </summary>
+    internal class Base64PipeReader : PipeReader
+    {
+        private readonly PipeReader _inner;
+        private ReadOnlySequence<byte> _currentInnerBuffer;
+        private ReadOnlySequence<byte> _currentDecodedBuffer;
+        private byte[]? _rentedBuffer;
+
+        public Base64PipeReader(PipeReader inner)
+        {
+            _inner = inner;
+        }
+
+        public override void AdvanceTo(SequencePosition consumed)
+        {
+            var consumedPosition = ResolvePosition(consumed);
+
+            _inner.AdvanceTo(consumedPosition);
+
+            ReturnBuffer();
+        }
+
+        public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
+        {
+            var consumedPosition = ResolvePosition(consumed);
+
+            // If the decoded buffer has no data then this is a cancelled result with some data.
+            // Report that we've examined to the end.
+            var examinedPosition = _currentDecodedBuffer.Length == 0
+                ? _currentInnerBuffer.End
+                : ResolvePosition(examined);
+
+            _inner.AdvanceTo(consumedPosition, examinedPosition);
+
+            ReturnBuffer();
+        }
+
+        private SequencePosition ResolvePosition(SequencePosition base64Position)
+        {
+            // This isn't ideal but it is the standard way to get the length between SequencePositions
+            var length = _currentDecodedBuffer.Slice(0, base64Position).Length;
+
+            // IMPORTANT - this logic assumes that SequencePosition will be multiples of 3.
+            // GetMaxEncodedToUtf8Length will round up to next multiple of 3, so more data
+            // could be advanced internally than expected if not called using a mulitple of 3.
+            //
+            // Because ReadAsync always returns data in multiples of 3, and gRPC only uses
+            // ReadResult.Buffer.End to advance, the position should always be a multiple
+            // of 3. The only time it won't be is at the end of a message, in which case
+            // rounding up is the correct thing to do to consume padding.
```

Comment:
I'm resubmitting this comment, since the last one was on a resolved conversation which I cannot unresolve.

This Base64PipeReader will be exposed to developers as `HttpContext.Request.BodyReader`, right? This would mean that developers can pass in arbitrary `examined` SequencePositions to PipeReader.AdvanceTo. In practice it makes sense that examined will always be `ReadResult.Buffer.End`, but we cannot know that for sure.

I still think we should throw if `length % 3 != 0` and we aren't at the end of the message.

Reply:
The pipereader doesn't know it is at the end of the message, but when it is then there will be padding, e.g. `(byte)'='`. I've added some validation.